libsplinter/src/admin/service/mod.rs

```diff
@@ -360,6 +335,32 @@ impl Service for AdminService {
             .add_services_to_directory()
             .map_err(|err| ServiceStartError::Internal(Box::new(err)))?;
 
+        let peer_admin_shared = self.admin_service_shared.clone();
+        let mut subscriber = self
+            .subscriber
+            .take()
+            .ok_or_else(|| ServiceStartError::AlreadyStarted)?;
+
+        thread::Builder::new()
+            .name("PeerInterconnect Receiver".into())
+            .spawn(move || loop {
+                let notification = match subscriber.next() {
+                    Some(notification) => notification,
+                    None => {
+                        warn!("PeerManager has shutdown");
+                        break;
+                    }
+                };
+
+                if let Ok(admin_shared) = peer_admin_shared.lock() {
+                    handle_peer_manager_notification(notification, admin_shared);
+                } else {
+                    error!("the admin shared lock was poisoned");
+                    break;
+                }
+            })
+            .map_err(|err| ServiceStartError::Internal(Box::new(err)))?;
```

Comment:
This thread should probably get joined on shutdown.

Reply:
update to return join handle from the admin service. Also updated the shutdown in splinterd to shutdown cleanly.